### DISCUSSION TITLE
Get instance of Store with generics from InstanceKeeper in right way

### DISCUSSION THREAD

**Discussion by t1r** [arkivanov/Essenty#96] [Q&A] [answered]
In short:
If we have two implementation of store interface with generic and take it from instance keeper, what is right way to get concrete store by key?
Just hardcode string, int or smth or something like Implementation1::class, Implementation2::class

More explanation of problem:
I have screen with 2 tabs. Every tab has a component implementation of the interface  like this:
```
interface AddFoodEntryComponent<T> {
    val models: Flow<Model<T>>
    val events: Flow<Event>
	//...
}
```

Every component has the store implementation of interface like this (with RecipeItem and ProductItem type in generic):
```
internal interface AddFoodEntryStore<T> : Store<Intent, State<T>, Label> {
    data class State<T>(
        val productList: List<T>? = null,
		//...
    )
	//...
}
```

We get instance of store like this:
```
private val store = instanceKeeper.getStore(!_KEY_!) { //If no key - Exception
	AddFoodEntryStoreProvider(...).provide()
}
```

If we don't provide key in `getStore(key)` function we will take Exception:
`java.lang.ClassCastException: RecipeItem cannot be cast to ProductItem`
I think it’s because Java generics erasure types mechanism.

Usage code:
```
// Implementation1
class AddFoodEntryStoreProvider() {
  fun provide(): AddFoodEntryStore<ProductItem> =
     object : AddFoodEntryStore<ProductItem>, Store<Intent, State<ProductItem>...
}

// Implementation2
class AddRecipeEntryStoreProvider() {
  fun provide(): AddFoodEntryStore<RecipeItem> =
     object : AddFoodEntryStore<RecipeItem>, Store<Intent, State<RecipeItem>...
}

class AddFoodEntryComponentImpl() 
  : AddFoodEntryComponent<ProductShort>, ComponentContext by componentContext {
    private val store = instanceKeeper.getStore(!_KEY_!) { //If no key - Exception
      AddFoodEntryStoreProvider(...).provide()
  }
}

class AddRecipeEntryComponentImpl() 
  : AddFoodEntryComponent<ProductShort>, ComponentContext by componentContext {
    private val store = instanceKeeper.getStore(!_KEY_!) { //If no key - Exception
      AddRecipeEntryStoreProvider(...).provide()
  }
}

AddFoodContainerComponentImpl() {
 constructor(...) : this(
        addFoodEntry = { childContext, params ->
            AddFoodEntryComponentImpl(
			  //...
            )
        },
        addRecipeEntry = { childContext, params ->
            AddRecipeEntryComponentImpl(
			  //...
            )
        },
 )
}
```

**Comment by arkivanov** [comment]
Thanks for the answer. If you have only one `Store` per child component, then everything should work automatically without specifying the key. Please make sure you pass the dedicated child `ComponentContext` into every child component (the one provided by the `childFactory` function), and not the parent's `ComponentContext`. `InstanceKeeper`'s keys must be unique within each instance of `ComponentContext`.

**Comment by t1r** [comment]
In my case it doesn't work automatically :)
I made sample with case from my question.
Is it ok pass StoreClass::class or Component::class to key in this case? Or mb exists better way?

Component 1
https://github.com/t1r/DecomposeSampleApp/blob/master/composeApp/src/commonMain/kotlin/org/sample/logic/food/integrationproduct/AddFoodEntryComponentImpl.kt

```
// Is class as key ok?
// Exception  without key
private val store = instanceKeeper.getStore(AddFoodEntryStoreProvider::class) {
  AddFoodEntryStoreProvider(
    storeFactory = storeFactory,
    foodRepository = foodRepository,
  ).provide()
}
```

Component 2
https://github.com/t1r/DecomposeSampleApp/blob/master/composeApp/src/commonMain/kotlin/org/sample/logic/food/integrationrecipe/AddRecipeEntryComponentImpl.kt
```
// Is class as key ok?
// Exception  without key
private val store = instanceKeeper.getStore(AddRecipeEntryStoreProvider::class) {
  AddRecipeEntryStoreProvider(
    storeFactory = storeFactory,
    foodRepository = foodRepository,
  ).provide()
}
```

**Comment by arkivanov** [comment, marked as answer]
I checked the sample, it doesn't work because you are creating both components with the same ComponentContext [here](https://github.com/t1r/DecomposeSampleApp/blob/53fe6df91c5e3902a4d9e9138634a1c5ac64eef2/composeApp/src/commonMain/kotlin/org/sample/logic/root/RootComponentImpl.kt#L59). As I mentioned earlier, sharing component contexts between multiple components is error prone and discouraged.

There are two options:

1. Wrap those two components into another container component, and use `childContext` for every child component.
2. Just write `addFoodEntry(componentContext.childContext("food"))`, and similarity for the second component.

In this case you won't need specifying keys in child components.

**Reply by t1r** [reply]
Thanks. It's clear now.
